File: functions/prueba.java
package com.steponeit.dentalone.appointments.services;

import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Isolation;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;
import java.util.stream.Collectors;

@Component
@Slf4j
public class TurnEaterService {
    protected Integer servers;
    protected List<Customer> customers;
    protected Date systemTime;

    @Transactional(isolation = Isolation.READ_COMMITTED)
    public void updateQueue() {
        List<Customer> beingServed = customers.stream().filter(Customer::isBeingServed).collect(Collectors.toList());
        List<Customer> waiting = customers.stream().filter(c -> !c.isBeingServed()).collect(Collectors.toList());
        updateBeingServed(beingServed);

        List<Customer> back = extract(waiting, servers);
        log.info("en servicio {}, espera {}, back {}", beingServed.size(), waiting.size(), back.size());
        waiting.removeAll(back);
        List<Customer> front = beingServed;

        updateWaiting(back, front);

        while (!waiting.isEmpty()) {
            //update back
            front = back;
            back = extract(waiting, servers);
            waiting.removeAll(back);
            log.info("En espera, {}", waiting.size());
            updateWaiting(back, front);
        }
    }

    private void updateBeingServed(List<Customer> beingServed) {
        for (Customer c : beingServed) {
            c.setCalculatedQueuePosition(0);
            c.setCalculatedTime(c.estimatedTimeLeft());
            //update served db
        }
    }

    private void updateWaiting(List<Customer> waiting, List<Customer> inFront) {
        for (Customer c : waiting) {
            Customer next = popLowest(inFront);
            c.setCalculatedQueuePosition(next.getCalculatedQueuePosition() + 1);
            c.setCalculatedTime(c.estimatedTimeLeft() + next.getCalculatedTime());
            c.setBehindName(next.getName());
        }
    }

    private Customer popLowest(List<Customer> customers) {
        if (customers.isEmpty()) {
            return null;
        }
        Optional<Customer> customerOptional = customers.stream()
                .sorted(Comparator.comparingLong(Customer::getCalculatedTime))
                .findFirst();
        if (customerOptional.isPresent()) {
            Customer customer = customerOptional.get();
            customers.remove(customer);
            return customer;
        } else {
            return null;
        }
    }

    public Integer getServers() {
        return servers;
    }

    public List<Customer> getCustomersInQueue() {
        return customers;
    }

    public static List<Customer> extract(List<Customer> customers, Integer servers) {
        List<Customer> ilTriello = new ArrayList<>();
        for (int i = 0; i < customers.size() && i < servers; i++) {
            ilTriello.add(customers.get(i));
        }
        return ilTriello;
    }

    public Customer createCustomer(String name, Date startedService, long calculatedTime) {
        return new Customer(name, calculatedTime, startedService);
    }

    @Data
    public class Customer {
        private final String name;
        private final Long servicesTime;
        private Date dateEntered;
        private Date dateStartedService;
        private Integer calculatedQueuePosition;
        private Long calculatedTime;
        private boolean beingServed = false;
        private String behindName;

        public Customer(String name, Long servicesTime, Date dateStartedService) {
            this.name = name;
            this.servicesTime = servicesTime;
            this.dateStartedService = dateStartedService;
        }

        public Long estimatedTimeLeft() {
            if (dateStartedService == null) {
                return servicesTime;
            } else {
                return servicesTime - (systemTime.getTime() - dateStartedService.getTime()) / 1000;
            }
        }
    }
}

File: functions/prueba.js
function crearCustomers(id_sucursal, horaInicioServ, tiempoCalculado) {
    var customers = [];
    const turnos = firebase.database().ref("/turno");
    const usuarios = firebase.database().ref("/usuario");

    let listaUsuario = {};
    usuarios.on("value", snapshot => {
        snapshot.forEach(function(childSnapshot) {
            let keys = childSnapshot.key;
            listaUsuario = Object.assign({
                    [keys]: childSnapshot.val()
                },
                listaUsuario
            );
        });
    });

    turnos.on("value", snapshot => {
        snapshot.forEach(function(childSnapshot) {
            let keys = childSnapshot.key;
            if (
                childSnapshot.child("id_sucursal").val() === id_sucursal &&
                childSnapshot.child("activo").val() === "true"
            ) {
                customers.push({
                    [childSnapshot.child("hora_pedido").val()]: {
                        ["id_turno"]: keys,
                        ["name"]: listaUsuario[childSnapshot.child("id_usuario").val()]["nombre"],
                        ["beingServed"]: false,
                        ["behindName"]: "",
                        ["dateStartedService"]: horaInicioServ,
                        ["calculatedTime"]: "",
                        ["calculatedQueuePosition"]: "",
                        ["servicesTime"]: tiempoCalculado
                    }
                });
            }
        });
    });
}

function estimatedTimeLeft(customer) {
    const fecha = new Date();
    if (customer["dateStartedService"].val() === null) {
        return customer["servicesTime"].val();
    } else {
        return (
            customer["servicesTime"].val() -
            (fecha.getTime() - customer["dateStartedService"].val()) / 1000
        );
    }
}

function extract(servers) {
    let ilTriello = [];
    for (i = 0; i < waiting.length && i < servers; i++) {
        ilTriello.push(waiting[i]);
    }
    return ilTriello;
}

function updateQueue(servers) {
    var beingServed = customers.filter(c => c["beingServed"].val());
    var waiting = customers.filter(c => !c["beingServed"].val());
    updateBeingServed();

    let back = extract(waiting, servers);
    back.forEach(item => {
        waiting.splice(waiting.indexOf(item), 1);
    });
    var front = beingServed;

    updateWaiting(back, front);

    while (!(waiting.length === 0)) {
        front = back;
        back = extract(servers);
        back.forEach(item => {
            waiting.splice(waiting.indexOf(item), 1);
        });
        updateWaiting(back, front);
    }
}

function updateBeingServed() {
    beingServed.forEach(c => {
        c["calculatedQueuePosition"] = 0;
        c["calculatedTime"] = estimatedTimeLeft(c);
    });
}

function updateWaiting() {
    waiting.forEach(item => {
        let next = popLowest();
        item["calculatedQueuePosition"] = next["calculatedQueuePosition"] + 1;
        item["calculatedTime"] = estimatedTimeLeft(item) + next["calculatedTime"];
        item["behindName"] = next["name"].val();
    });
}

function popLowest() {
    if (front.length <= 0) {
        return null;
    }
    let menor = 999999;
    let customer = null;

    front.forEach(item => {
        if (item["calculatedTime"].val() < menor) {
            menor = item["calculatedTime"].val();
            customer = item;
        }
    });
    front.splice(waiting.indexOf(customer), 1);
    return customer;
}
var servers = 2;
let cont = 0;
var customers = [];
customers.push({
    [cont++]: {
        ["id_turno"]: "15256",
        ["name"]: "a",
        ["beingServed"]: false,
        ["behindName"]: "",
        ["dateStartedService"]: null,
        ["calculatedTime"]: "",
        ["calculatedQueuePosition"]: "",
        ["servicesTime"]: 200
    }
});

customers.push({
    [cont++]: {
        ["id_turno"]: "1556",
        ["name"]: "b",
        ["beingServed"]: false,
        ["behindName"]: "",
        ["dateStartedService"]: null,
        ["calculatedTime"]: "",
        ["calculatedQueuePosition"]: "",
        ["servicesTime"]: 100
    }
});

customers.push({
    [cont++]: {
        ["id_turno"]: "152996",
        ["name"]: "c",
        ["beingServed"]: false,
        ["behindName"]: "",
        ["dateStartedService"]: null,
        ["calculatedTime"]: "",
        ["calculatedQueuePosition"]: "",
        ["servicesTime"]: 100
    }
});

customers.push({
    [cont++]: {
        ["id_turno"]: "152696",
        ["name"]: "d",
        ["beingServed"]: false,
        ["behindName"]: "",
        ["dateStartedService"]: null,
        ["calculatedTime"]: "",
        ["calculatedQueuePosition"]: "",
        ["servicesTime"]: 50
    }
});

customers.push({
    [cont++]: {
        ["id_turno"]: "15256ff",
        ["name"]: "e",
        ["beingServed"]: false,
        ["behindName"]: "",
        ["dateStartedService"]: null,
        ["calculatedTime"]: "",
        ["calculatedQueuePosition"]: "",
        ["servicesTime"]: 70
    }
});

updateQueue(servers);
customers.forEach(item => {
    console.log(
        `customer:${item["name"].val()}  
        ${item["calculatedQueuePosition"].val()}
        ${item["calculatedTime"].val()} 
        ${item["behindName"].val()}`
    );
});